### ISSUE TITLE
Driver spins up secondary thread

### ISSUE THREAD

**Issue by Sheado** [glfw/glfw#595]
Hi All, 

I've been trying to chase down this issue for a while now - posting in different places with no luck:
nvidia forums: https://devtalk.nvidia.com/default/topic/832506/opengl/nvoglv32-consuming-a-ton-of-cpu/
stackoverflow: http://stackoverflow.com/questions/30387961/occasional-high-cpu-usage-by-nvoglv32-dll-threads-in-opengl-win32-application

I can't figure out if it's a driver issue or if it's a GLFW issue. I'm pretty sure it's not my code because I was able to reproduce it using the GLFW sample (as described in the posts linked above). To summarize, on occasion (1 in 3-4 times) my GLFW game consumes a ton of cpu. Other times, it runs as expected and doesn't consume much cpu. The main difference between these two cases is that when it's running inefficiently nvoglv32.dll spawns an extra thread. Note: I'm testing on Windows 8 & 10 with NVIDIA GeForce GT 640M 512 MB.

The threads seem to get created from the GLFW code after the call to:
`
nativeCount = DescribePixelFormat(window->wgl.dc,
                                          1,
                                          sizeof(PIXELFORMATDESCRIPTOR),
                                          NULL);
`
Anybody experience this problem? Any ideas on how to solve it? Help!

Thanks

**Comment by hypernewbie** [comment]
Does forcing threaded-optimisation off from the control panel do anything?

**Comment by Sheado** [comment]
Thanks for the response hypernewbie, 
I'll try that out on Monday and get back to you.

**Comment by hypernewbie** [comment]
My wild guess here would be that the driver (which heuristically decides whether to use threaded optimisation or not, if none specified) may spawn an extra thread busy-spinning and submitting commands from the command-producing threads. This is less overhead than continually waiting on a kernel lock, especially if there is nothing else being scheduled on that kthread anyway.

**Comment by Sheado** [comment]
Interesting theory - makes sense, but wouldn't they put some sleeps in their busy spins? 
Hopefully turning off threaded-optimization fixes it for me - that would be awesome! I'll respond as soon as I can.

**Comment by Sheado** [comment]
Hey hypernewbie,

So far it seems like you're right! From my testing:
Setting "threaded optimization":
* off - issue did not occur - tested 40 times
* on - occurred within 10 tries
* auto - occurred within 20 tries

I need to test more to make sure, but this seems to be related to threaded optimization. If this is the case, then my questions are:
 * Is there any way to request no threaded optimizations from GLFW code? Asking users to change their settings is always a difficult task.
 * Is this a bug in GLFW? Why do some games benefit from the threaded optimization setting, while GLFW not so much.

Any help is much appreciated!

**Comment by theagentd** [comment]
Some GPU commands, specifically buffer mapping and readbacks of data, force the application and driver thread to synchronize. In this case, threading reduces performance:

Application thread queues up commands, driver thread is waiting for commands. Application thread maps a buffer and needs to wait for driver thread. Driver thread starts working while application thread waits.

The driver tries to detect that using heuristics and disables threading, but it doesn't always get it right.

**Comment by Sheado** [comment]
Thanks for the explanation theagentd.
Speaking of synchronization, the control panel mentions that threaded optimization kicks in when vsync is on. Does that mean turning off vsync, turns off threaded optimization?

**Comment by theagentd** [comment]
No, V-sync has nothing to do with threaded optimization. It just improves CPU performance on multi-core processors by offloading the majority of the driver's work to a different thread. With good coding, it can improve performance quite a bit by cutting down the time your app spends on inherently single-threaded OpenGL calls. Intel and AMD drivers do the exact same thing, but don't have an option to enable or disable it as far as I know.

**Comment by Sheado** [comment]
That's too bad.. I was hoping that the v-sync setting affects the threaded optimization heuristic.

**Comment by theagentd** [comment]
It probably does affect the heuristic's ability to determine whether to enable or disable it when set to Auto since it caps the frame rate, but not in a predictable way. Sorry. xd

What exactly is your problem here? Do you want to force it to on or off? Why?

**Comment by Sheado** [comment]
Ah, ok, then I'll give it a shot when I get a chance =]

I want to try to force it off when it is set to auto. 

In testing my game and the GLFW samples, an extra nvoglv32.dll thread is occasionally spawning and with that comes a sluggish frame rate. When that extra thread doesn't spawn, performance is good. Also, after testing with "threaded optimization" set to OFF I wasn't able to reproduce the problem in the 40 tests I ran. So I'm hoping to either find a way to prevent threaded optimization from triggering or better yet a solution that makes GLFW consistently benefit from threaded optimization.

**Comment by elmindreda** [comment]
GLFW has nothing to do with this. You're talking directly to the driver ICD.

**Comment by elmindreda** [comment]
@Sheado If you want to talk to the driver about settings, have a look at [NVAPI](https://developer.nvidia.com/nvapi).

**Comment by Sheado** [comment]
Hi elmindreda,

Thanks for the link - I'll take a look - this could be a good last resort for me.
Any ideas as to why I'm experiencing occasional sluggishness when threaded optimization kicks in? 

Note: the way I'm testing this is by simply launching my game or the sample app. When things are running fine for my game CPU usage is under 20%. On a launch when that extra thread kicks in, CPU goes over 50%.

**Comment by theagentd** [comment]
Is it affecting performance at all? I believe Nvidia's driver sometimes implements waiting as yield looping, which means it won't compete with other threads about CPU time.

**Comment by Sheado** [comment]
@theagentd - will do. I'm busy preparing an alpha test at the moment, so my responses my slow down for a couple of weeks while I deal with that. I'll respond with profiling details as soon as I can. Last time I profiled it I remember most of the work being done in nvoglv32.dll, but I don't remember what functions were causing that. I do remember it being difficult to profile because it could only reproduce it once every few tries.

**Comment by theagentd** [comment]
What do you mean? If manually setting Threaded Optimization to true doesn't reproduce the issue 100% of the time, then it's most likely something else. In addition, you want to figure out exactly which OpenGL function is taking so much time. If you want, you can add me on Skype (same username as here).

**Comment by Sheado** [comment]
Thanks for the help theagentd, 
Added you on Skype.

To clarify, Threaded Optimization set to OFF so far does not reproduce the issue (I tested 40 times and plan to test more to verify). With it ON or AUTO it always happens within 10 tries. 

I hope to have more info in a couple weeks after we're done with the alpha test of our game.

**Comment by hypernewbie** [comment]
Sheado: I think you are overreacting here.

The 2nd thread takes a lot of CPU, but on another core, not affecting your main program much. The FPS lower you're seeing is probably from threading overhead. Threaded command submission has higher overhead (obviously) but has other benefits of its own; the driver usually has some sort of heuristics to determine whether the overhead is worth the benefits. The heuristics doesn't always decide the right thing all the time; thats why its called heuristics :)

As dreda mentioned, for manual control, you can use the NVAPI interface to turn threaded-optimisation off for your app, and avoid this whole driver heuristics business. Yes, Intel / AMD do do similar things; and yeah no publically exposed options to control them.

**Comment by Sheado** [comment]
hi hypernewbie,

When the problem happens my FPS is dropping from a steady 60fps all the way down to ~25fps. I'd like to find a solution in order to avoid negative reviews from gamers who happen to have a machine config similar to the one I'm experiencing difficulties with. 

My next plan of action is:
1. Provide a vsync on/off option in hopes of it affecting the heuristic in my favor when the issue happens. 
2. If that doesn't work, then tinker with NVAPI to shut off threaded optimizations for my game.

**Comment by hypernewbie** [comment]
Hmm, now that does sounds like a perf regression of some sort. Threaded optimisation overhead should not get this bad. I'll have a look with a CPU profiler in the next week or so with driver symbols. 

Do you have an email I can contact you via, so we don't send more spam to the GLFW guys with this?

dreda: Let's call this one a close, it's likely unrelated to GLFW at all.

**Comment by theagentd** [comment]
I have also written programs that suffer immensely from threaded optimization (that didn't use GLFW). It's not uncommon. Again, my tip is to code your program in a way that it can take advantage of the threading.

**Comment by Sheado** [comment]
Thanks for all the help everybody, 
Sorry if I spammed you all with this issue. 

@hypernewbie - You can contact me at "chad at sheado.net"

**Comment by Sheado** [comment]
Hi All, 

I'm still struggling with this issue (and game is shipping in 3 months!) and I just wanted to update with my latest findings in case anybody else has come across a solution. I know this is probably not a GLFW issue, but the workaround might be in how I'm using GLFW. 

The latest thing I've noticed is that if I remove joystick polling, my game never lags - fps is always at 60. So basically removing the following lines of code from my main loop helps (single threaded - called before drawing calls happen):
	
    const unsigned char* values = glfwGetJoystickButtons(id, &count);
    const float* axes = glfwGetJoystickAxes(id, &count);

I measured the time it takes for this code to run and it's only 1-2ms. So the polling itself isn't cutting the FPS in half (which is what happens when the threaded optimization issue happens). 

Can anybody think of why this would cause a conflict with threaded optimization? Anybody have an idea of how I can keep joystick support while not experiencing the performance degradation associated with threaded optimization?

Thanks!
